packages/block-editor/src/components/list-view/block-rename-ui.js

```diff
@@ -0,0 +1,136 @@
+import { useState, forwardRef, useEffect } from '@wordpress/element';
+import { ENTER, ESCAPE } from '@wordpress/keycodes';
+import { __, sprintf } from '@wordpress/i18n';
+
+const ListViewBlockRenameUI = forwardRef(
+	( { blockTitle, onSubmit, onCancel }, ref ) => {
+		const inputRef = useFocusOnMount();
+		const inputDescriptionId = useInstanceId(
+			ListViewBlockRenameUI,
+			`block-editor-list-view-block-node__input-description`
+		);
+
+		const formDescriptionId = useInstanceId(
+			ListViewBlockRenameUI,
+			`block-editor-list-view-block-node__form-description`
+		);
+
+		// Local state for value of input **pre-submission**.
+		const [ inputValue, setInputValue ] = useState( blockTitle );
+
+		const onKeyDownHandler = ( event ) => {
+			// Trap events to input when editing to avoid
+			// default list view key handing (e.g. arrow
+			// keys for navigation).
+			event.stopPropagation();
+
+			// Handle ENTER and ESC exits editing mode.
+			if ( event.keyCode === ENTER || event.keyCode === ESCAPE ) {
+				if ( event.keyCode === ESCAPE ) {
+					handleCancel();
+				}
+
+				if ( event.keyCode === ENTER ) {
+					handleSubmit();
+				}
+			}
+		};
+
+		const handleCancel = () => {
+			// Reset the input's local state to avoid
+			// stale values.
+			setInputValue( blockTitle );
+
+			onCancel();
+
+			// Must be assertive to immediately announce change.
+			speak( 'Leaving block name edit mode', 'assertive' );
+		};
+
+		const handleSubmit = () => {
+			// Submit changes only for ENTER.
+			onSubmit( inputValue );
+
+			let successAnnouncement;
+
+			if ( inputValue === '' ) {
+				successAnnouncement = __( 'Block name reset.' );
+			} else {
+				successAnnouncement = sprintf(
+					/* translators: %s: new name/label for the block */
+					__( 'Block name updated to: "%s".' ),
+					inputValue
+				);
+			}
+
+			// Must be assertive to immediately announce change.
+			speak( successAnnouncement, 'assertive' );
+		};
+
+		useEffect( () => {
+			speak( 'Entering block name edit mode', 'assertive' );
+		}, [] );
+
+		return (
+			<Popover
+				anchorRef={ ref }
+				placement={ 'overlay' }
+				resize={ true }
+				variant={ 'unstyled' }
+				animate={ false }
+				className="block-editor-list-view-block-node__rename-popover"
+			>
+				<form
+					className="block-editor-list-view-block-node__input"
+					onSubmit={ ( e ) => {
+						e.preventDefault();
+
+						onSubmit( inputValue );
+					} }
+					aria-describedby={ formDescriptionId }
+				>
+					<VisuallyHidden>
+						<p id={ formDescriptionId }>
+							{ __( 'Choose a custom name for this block.' ) }
+						</p>
+					</VisuallyHidden>
+					<InputControl
+						ref={ inputRef }
+						value={ inputValue }
+						label={ __( 'Edit block name' ) }
+						hideLabelFromVision
+						onChange={ ( nextValue ) => {
+							setInputValue( nextValue ?? '' );
+						} }
+						onBlur={ () => {
+							// Cancel editing mode.
+							handleCancel();
+						} }
```

Comment:
As blur will stop editing do we need to make some kind of focus trap to ensure the user cannot tab out of the interface without explicitly dismissing it?

Reply:
At this point, it's starting to sound a lot like a modal. Can it be a modal, but visually styled to be inline on the location where the input would appear? There's no rule AFAIK that modals have to be full screen overlays.

Reply:
It doesn't have to be a visual modal, but it needs to either use `<modal>` tag or `role="modal"`. Even if constrained tabbing pattern is used, this does not prevent screen readers virtual mode from navigating out of the popover. Modals are the only element I know of that can do this.

Reply:
Noted with thanks. I was wondering whether it would need to be a full modal.